### PR TITLE
uncomment shared file read and write

### PR DESCRIPTION
**Pull request by kevodwyer**

### CODE REVIEW

File: src/peergos/server/simulation/PeergosFileSystemImpl.java [hunk 1]
```diff
@@ -50,6 +54,10 @@ private FileWrapper getDirectory(Path path) {
     @Override
     public void write(Path path, byte[] data, Consumer<Long> progressConsumer) {
         FileWrapper directory = getDirectory(path);
+        Optional<FileWrapper> existingFile = userContext.getByPath(path).join();
+        if (existingFile.isPresent() && existingFile.get().getFileProperties().isDirectory) {
+            return;
```
Comment (ianopolous) [inline src/peergos/server/simulation/PeergosFileSystemImpl.java:59]: Couldn't you check it is writable?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 2]
```diff
@@ -515,16 +543,26 @@ private boolean verifySharingPermissions(String user, Path path) {
...
-                                fs.write(path, new byte[]{read[0]});
-                            } catch (Exception ex) {
-                                LOG.log(Level.SEVERE, "User " + sharee + " could not write  shared-path " + path + "!", ex);
-                                isVerified = false;
+                                fs.write(path, read == null ? null : new byte[]{read[0]});
+                            } catch (Throwable ex) {
+                                //reload user and try again
+                                fs = fileSystems.getTestFileSystem(sharee, true);
+                                try {
+                                    fs.write(path, read == null ? null : new byte[]{read[0]});
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java:551]: why is this necessary?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 3]
```diff
@@ -93,8 +93,8 @@ public void addUser(String user) {
     enum Simulation {
         READ_OWN_FILE,
         WRITE_OWN_FILE,
-        READ_SHARED_FILE,
-        WRITE_SHARED_FILE,
+        READ_SHARED_FILE_OR_DIRECTORY,
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java:96]: Do you think these should be two different things? reading a file is very different to reading a directory. The latter is doing a directory listing.

File: src/peergos/server/tests/simulation/Simulator.java [hunk 4]
```diff
@@ -515,16 +543,26 @@ private boolean verifySharingPermissions(String user, Path path) {
                         if (isVerified) {
                             try {
                                 // can overwrite?
-                                fs.write(path, new byte[]{read[0]});
-                            } catch (Exception ex) {
-                                LOG.log(Level.SEVERE, "User " + sharee + " could not write  shared-path " + path + "!", ex);
-                                isVerified = false;
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:521]: Can you explain what the logic for this is?

File: src/peergos/server/tests/simulation/Simulator.java (hunk 3 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:96]: I think they should be different operations

File: src/peergos/server/tests/simulation/NativeFileSystemImpl.java [hunk 5]
```diff
@@ -70,7 +70,9 @@ private void ensureCan(Path path, Permission permission, String user) {
     public byte[] read(Path path, BiConsumer<Long, Long> pc) {
         Path nativePath = virtualToNative(path);
         ensureCan(path, Permission.READ);
-
+        if (nativePath.toFile().isDirectory()) {
+           return null;
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/NativeFileSystemImpl.java:74]: This would be easier to reason about if it raised an exception. Why return null?

File: src/peergos/server/simulation/PeergosFileSystemImpl.java [hunk 6]
```diff
@@ -42,6 +43,9 @@ private FileWrapper getDirectory(Path path) {
     public byte[] read(Path path, BiConsumer<Long, Long> progressConsumer) {
         FileWrapper wrapper = getPath(path);
         long size = wrapper.getFileProperties().size;
+        if (wrapper.getFileProperties().isDirectory) {
+            return null;
```
Comment (cboddy) [inline src/peergos/server/simulation/PeergosFileSystemImpl.java:47]: This should raise

File: src/peergos/server/simulation/PeergosFileSystemImpl.java (hunk 1 above)
Comment (kevodwyer) [reply src/peergos/server/simulation/PeergosFileSystemImpl.java:59]: sure, also see answer below

File: src/peergos/server/tests/simulation/Simulator.java (hunk 2 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:551]: That is the question. I don't know the answer at this stage. I am investigating. Making this change was sufficient for verify() to succeed.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 3 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:96]: ok, I was attempting to make the least amount of changes (no changes to interfaces). The thought being reading/writing actual file is only one step further past checking existence of file/directory.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 4 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:521]: working on it. The above solves the problem, but why the snapshot is out-of-date is not answered yet.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:551]: Maybe check it's not using a caching mutable pointers?

File: src/peergos/server/tests/simulation/NativeFileSystemImpl.java (hunk 5 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/NativeFileSystemImpl.java:74]: The method still confirms the file exists and has the right permissions. I didn't want to add read/write directory specific method to interface. But ok, message heard. I can do so.

File: src/peergos/server/simulation/PeergosFileSystemImpl.java (hunk 6 above)
Comment (kevodwyer) [reply src/peergos/server/simulation/PeergosFileSystemImpl.java:47]: same as above. I can revisit.

File: src/peergos/server/simulation/PeergosFileSystemImpl.java (hunk 1 above)
Comment (kevodwyer) [reply src/peergos/server/simulation/PeergosFileSystemImpl.java:62]: fixed

File: src/peergos/server/tests/simulation/Simulator.java (hunk 2 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java]: grant code did not take into account any previous call to grant

File: src/peergos/server/tests/simulation/Simulator.java (hunk 3 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java]: see new code

File: src/peergos/server/tests/simulation/NativeFileSystemImpl.java (hunk 5 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/NativeFileSystemImpl.java]: done now

File: src/peergos/server/simulation/PeergosFileSystemImpl.java (hunk 6 above)
Comment (kevodwyer) [reply src/peergos/server/simulation/PeergosFileSystemImpl.java]: separated file and directory cases now

File: src/peergos/server/tests/simulation/Simulator.java (hunk 4 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:521]: removed - was due to simulator not taking into account existing sharing state

**Comment by kevodwyer** [comment]
have updated PR to take into account review

**Comment by kevodwyer** [comment]
I have been running it all afternoon. I have gone back to global verify after xyz operations else i run out of file descriptors on my machine.

**Comment by kevodwyer** [comment]
I have been running it all afternoon. I have gone back to global verify after xyz operations else i run out of file descriptors on my machine.

**Comment by kevodwyer** [comment]
PR  open until i sort out a rare bug

**Comment by kevodwyer** [comment]
bug was not in simulator. this PR is ready to be reviewed

**Comment by cboddy** [comment]
Sorry for missing the context but can you say what the objective of this change is, please?

**Comment by kevodwyer** [comment]
Implement READ_SHARED_FILE, READ_SHARED_DIRECTORY, WRITE_SHARED_FILE, WRITE_SHARED_DIRECTORY

File: src/peergos/server/tests/simulation/Simulator.java [hunk 7]
```diff
@@ -263,13 +270,18 @@ public Simulator(int opCount, Random random, long meanFileLength,
...
+    private void readFile(String user, Path path) {
         FileSystem testFileSystem = fileSystems.getTestFileSystem(user);
         FileSystem referenceFileSystem = fileSystems.getReferenceFileSystem(user);
+        referenceFileSystem.read(path);
+        testFileSystem.read(path);
+    }
 
-        byte[] refBytes = referenceFileSystem.read(path);
-        byte[] testBytes = testFileSystem.read(path);
-        return Arrays.equals(refBytes, testBytes);
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:272]: Why is  this not comparing the bytes anymore?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 8]
```diff
@@ -263,13 +270,18 @@ public Simulator(int opCount, Random random, long meanFileLength,
...
+    private void readDirectory(String user, Path path) {
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:280]: 1.  I think the concept of "reading" a  directory is a bit confusing  and doesn't directly map onto (is it read & execute?) - why introduce it?
2. What is this supposed to check?  Possible side effects of  `ls`-ing a  directory?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 9]
```diff
@@ -94,7 +94,9 @@ public void addUser(String user) {
         READ_OWN_FILE,
         WRITE_OWN_FILE,
         READ_SHARED_FILE,
+        READ_SHARED_DIRECTORY,
         WRITE_SHARED_FILE,
+        WRITE_SHARED_DIRECTORY,
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:99]: read means you can read the contents. write means you can add files or subdirs

[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:99]: This is what unix does.

File: src/peergos/server/tests/simulation/Simulator.java [hunk 10]
```diff
@@ -287,8 +299,10 @@ private void write(String user, Path path) {
         Path dirName = path.getParent();
         String fileName = path.getFileName().toString();
         Map<Path, List<String>> dirsToFiles = index.getDirToFiles(user);
-        dirsToFiles.get(dirName).add(fileName);
-
+        List<String> existingFiles = dirsToFiles.get(dirName);
+        if (! existingFiles.contains(fileName)) {
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:303]: this  makes me a bit nervous - should `existingFiles` be updated to a set instead?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 11]
```diff
@@ -301,6 +315,26 @@ private boolean grantPermission(String granter, String grantee, Path path, FileS
         FileSystem testFileSystem = fileSystems.getTestFileSystem(granter);
         FileSystem referenceFileSystem = fileSystems.getReferenceFileSystem(granter);
 
+        Set<String> testExistingWriters = new TreeSet<>(testFileSystem.getSharees(path, FileSystem.Permission.WRITE));
+        Set<String> refExistingWriters = new TreeSet<>(referenceFileSystem.getSharees(path, FileSystem.Permission.WRITE));
+        if (! testExistingWriters.toString().equals(refExistingWriters.toString())) {
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:320]: nit-pick: why convert to string when comparing?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 12]
```diff
@@ -376,35 +410,72 @@ private void run(final Simulation simulation, final String user) {
...
         switch (simulation) {
             case READ_OWN_FILE:
                 Path readPath = randomFilePathForUser.get();
                 log(user, simulation, readPath);
-                read(user, readPath);
+                readFile(user, readPath);
                 break;
             case READ_SHARED_FILE:
-                Optional<Path> sharedOpt = randomSharedPath.apply(otherUser.get(), user);
+                Optional<Path> sharedOpt = randomSharedFilePath.apply(user, otherUser.get());
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:443]: Formerly this would result in `user` reading a (random) file that another (random)  user had shared with `user` -  is that still the case?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 13]
```diff
@@ -376,35 +410,72 @@ private void run(final Simulation simulation, final String user) {
...
+                readDirectory(user, sharedDirPathToRead);
                 break;
             case WRITE_OWN_FILE:
                 Path path = getAvailableFilePath(user);
                 log(user, Simulation.WRITE_OWN_FILE, path);
                 write(user, path);
                 break;
             case WRITE_SHARED_FILE:
-                Optional<Path> sharedOpt2 = randomSharedPath.apply(otherUser.get(), user);
+                Optional<Path> sharedOpt2 = randomSharedFilePath.apply(user, otherUser.get());
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:464]: Same question as above

File: src/peergos/server/tests/simulation/Simulator.java [hunk 14]
```diff
@@ -376,35 +410,72 @@ private void run(final Simulation simulation, final String user) {
...
                 if (! sharedOpt2.isPresent())
                     return;
                 Path  sharedPathToWrite = sharedOpt2.get();
                 log(user, Simulation.WRITE_SHARED_FILE, sharedPathToWrite);
                 write(user, sharedPathToWrite);
                 break;
+            case WRITE_SHARED_DIRECTORY:
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:471]: Can we call this `MKDIR_SHARED` instead, please?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 15]
```diff
@@ -461,8 +526,13 @@ public void run() {
             System.out.println("iOp=" + iOp);
             Simulation simulation = getNextSimulation.get();
             run(simulation);
+            /*
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:529]: commented  code to be removed

File: src/peergos/server/tests/simulation/Simulator.java [hunk 16]
```diff
@@ -486,44 +558,79 @@ private boolean verifySharingPermissions(String user, Path path) {
                 LOG.info("User " + peergosFileSystem.user() + " path " + path + " has peergos-fs " + permission.name() + "ers " +
                         shareesInPeergos + " and local-fs " + permission.name() + "ers " + shareesInLocal);
                 isVerified = false;
+                break;
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:561]: Why break here?  Even if the permissions  are wrong we still want to check the contents  for  `user`/`path`.

File: src/peergos/server/tests/simulation/Simulator.java [hunk 17]
```diff
@@ -486,44 +558,79 @@ private boolean verifySharingPermissions(String user, Path path) {
...
-                        try {
-                            //can read?
-                            read = fs.read(path);
-                        } catch (Exception ex) {
-                            LOG.log(Level.SEVERE, "User "+ sharee +" could not read shared-path "+ path +"!", ex);
-                            isVerified = false;
-                        }
-                        if (isVerified) {
+                PeergosFileSystemImpl fs = fileSystems.getTestFileSystem(sharee);
+                if (isDir) {
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:568]: I  think forcing reading and writing  for  directories and files into two ops makes this quite complicated.

Suggestion: simpligy by removing the `READ_SHARED_DIR` and `WRITE_SHARED_DIR` ops and add new `MKDIR_SHARED` and  `LS_SHARED` operations that will create or list the contents  of  a (random) shared directory,  or raise an error if  path is a file.

File: src/peergos/server/tests/simulation/Simulator.java [hunk 18]
```diff
@@ -486,44 +558,79 @@ private boolean verifySharingPermissions(String user, Path path) {
...
+                            } catch (Exception ex) {
+                                LOG.log(Level.SEVERE, "User " + sharee + " could not read shared-path " + path + "!", ex);
+                                isVerified = false;
+                            }
+                            break;
+                        case WRITE:
                             try {
-                                // can overwrite?
-                                fs.write(path, new byte[]{read[0]});
+                                //can read?
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:612]: Isn't this no longer checking that the path can be written? It's only being read now.

File: src/peergos/server/tests/simulation/Simulator.java [hunk 19]
```diff
@@ -486,44 +558,79 @@ private boolean verifySharingPermissions(String user, Path path) {
...
+                                read = fs.read(path);
                             } catch (Exception ex) {
-                                LOG.log(Level.SEVERE, "User " + sharee + " could not write  shared-path " + path + "!", ex);
+                                LOG.log(Level.SEVERE, "User " + sharee + " could not read shared-path " + path + "!", ex);
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:615]: Can you make the log message  explicit that this should be a _writable_ file that failed to be read?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 20]
```diff
@@ -486,44 +558,79 @@ private boolean verifySharingPermissions(String user, Path path) {
...
                                 isVerified = false;
                             }
-                        }
-                        break;
-                    default:
-                        throw new IllegalStateException();
+                            if (isVerified) {
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:618]: Why only do  this if  the permissions are  correct? Also this way it won't log the problem (that the path is  not writable).

File: src/peergos/server/tests/simulation/Simulator.java (hunk 14 above)
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:471]: Write is the terminology unix uses. We are just using mkdir to test it

File: src/peergos/server/tests/simulation/Simulator.java [hunk 21]
```diff
@@ -535,13 +642,23 @@ private boolean verifySharingPermissions(String user, Path path) {
...
+        Stat stat = fileSystems.getReferenceFileSystem(user).stat(path);
+        boolean isDir = stat.fileProperties().isDirectory;
         try {
-            byte[] testData = peergosFileSystem.read(path);
-            byte[] refData = nativeFileSystem.read(path);
-            if (!Arrays.equals(testData, refData)) {
-                LOG.info("Path " + path + " has different contents between the file-systems");
-                return false;
+            if (isDir) {
+                Set<Path> refDirectoryListing = new TreeSet<>(nativeFileSystem.ls(path, true));
```
Comment (cboddy) [inline src/peergos/server/tests/simulation/Simulator.java:649]: I'm not sure this  is necessary -  by checking  each  _file_ in both filesystems the list of files in each directory is also implicitly checked.

What this will check is if there is some problem with the  result of the `ls` call in each directory - a view that should be the same as reading all the files we think are there. If we're going to check  `ls`  explicitly I think do  it separately - iterate through  all the directories and check the `ls` rather  than shoe-horning it in here.

Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:649]: You can be granted read access to things in a directory and not have full read access to the parent dir, this means you wouldn't be able to see siblings of the files. Similarly you can be granted write access to a file, but not even read to its parent. This means you can modify that file, but not rename it, nor ls the parent dir. This is what unix does.

Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:649]: Yes, and this should still be checked separately.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:99]: That depends on the execute bit  for that directory (in unix).

[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:99]: It is implemented as W && X, but that's not the semantic meaning, just an implementation detail: https://en.wikipedia.org/wiki/File_system_permissions

File: src/peergos/server/tests/simulation/Simulator.java (hunk 21 above)
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:649]: What about an empty directory? Just checking files won't catch that.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 7 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:272]: The return value was never checked previously anyway. The contents is now checked as part of verify()

File: src/peergos/server/tests/simulation/Simulator.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:99]: As discussed it's actually `rwx` right now

[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:99]: In unix speak yes, but semantically it's: R means you can list and read the children. W means you can add and rename children (and assumes you also have R)

File: src/peergos/server/tests/simulation/Simulator.java (hunk 8 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:280]: 1. read == list. we use the word read for file so it seemed appropriate. we have write directory for make directory.
2. yes, the code can throw exceptions under a number of circumstances.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 10 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java]: done

File: src/peergos/server/tests/simulation/Simulator.java (hunk 11 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:318]: force of habit, fixed

File: src/peergos/server/tests/simulation/Simulator.java (hunk 15 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:527]: 2 modes, verify after each op, or verify after all ops. the former useful when debugging, later is what it was originally

File: src/peergos/server/tests/simulation/Simulator.java (hunk 16 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:559]: if the permissions are wrong there is a bug. When bug is fixed then the extra processing is done.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 19 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:613]: fixed

File: src/peergos/server/tests/simulation/Simulator.java (hunk 20 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:616]: every time isVerfied is set to false a SEVERE log message is written. If that occurs, i expect the bug to be fixed and then the extra checks are done.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 16 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:559]: The point is we still want to know if the contents match if the permissions are wrong

File: src/peergos/server/tests/simulation/Simulator.java (hunk 12 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:441]: Did you have a chance to check this?

Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:441]: I'm still working through your comments. I get a fatal exception with the params in the original position. I hope to have time tomorrow to investigate further.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 15 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java:527]: Oh I see. Leo free to add a flag to do this too :)

You still need to check everything at the end in case of regressions that affect fonts/folders other than the path directly referenced in the simulation op.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 12 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java]: Reversing the order of params like i did originally meant i could compare ref and peergos filesystems. 
As ref file system does not use sym links to model the shared dir structure i can't compare between the filesystems. That's fine i now just check to make sure the fie/dir exists in peergos dir.

File: src/peergos/server/tests/simulation/Simulator.java (hunk 13 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:466]: answer above

File: src/peergos/server/tests/simulation/Simulator.java (hunk 16 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java]: done

File: src/peergos/server/tests/simulation/Simulator.java (hunk 17 above)
Comment (kevodwyer) [reply src/peergos/server/tests/simulation/Simulator.java:569]: I'm reluctant to change this again. It does the job and i think it is clear. Can be enhanced in a subsequent PR.

**Comment by kevodwyer** [comment]
@cboddy @ianopolous I think i have addressed all outstanding comments. Have run simulator again a number of times and all looks good.  There is a an item on the project board for further fuzzer work that can be done in another PR. https://github.com/Peergos/Peergos/issues/392

File: src/peergos/server/tests/simulation/Simulator.java (hunk 17 above)
Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java:569]: The fact that it is doing a mkdir and ls is just an implementation detail. Semantically it is testing whether we can read the directory, or write to the directory. And indeed we could change the implementation to for example delete the dir made. Instead of ls we could also check we have a read key in the Peergos case.

**Comment by ianopolous** [comment]
If no objections by tomorrow, we should merge this.
